### PR TITLE
txn: do conflict check before returning assertion failure

### LINKED ISSUE #12487: Do constraint check before returning an assertion error
## Development Task
ref: 
https://github.com/tikv/client-go/pull/492
https://github.com/pingcap/tidb/issues/33641

Assertion may be false-positive for a secondary index key.
Briefly the case is :
1. txn-a prewrites key-1
2. txn-b prewrites index-1
3. txn-b prewrites key-1, and encounters lock
4. txn-b rolls back txn-a
5. txn-b prewrites key-1, and successfully commit
6. txn-a prewrites index-1, and succeeds (because there is no constraint check) 

We enforce a constraint check before the assertion error is returned. If the constraint check failed, return it instead of the assertion failure.

### PR DESCRIPTION
**Pull request by ekexium**
Signed-off-by: ekexium <ekexium@gmail.com>

### What is changed and how it works?

Issue Number: Close #12487 

What's Changed:

We enforce a constraint check before the assertion error is returned. If the constraint check failed, return it instead of the assertion failure.

```commit-message
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/txn/commands/prewrite.rs [hunk 1]
```diff
@@ -2189,7 +2189,7 @@ mod tests {
         );
         assert!(matches!(
             err,
-            MvccError(box MvccErrorInner::AssertionFailed { .. })
+            MvccError(box MvccErrorInner::WriteConflict { .. })
```
Comment (sticnarf) [inline src/storage/txn/commands/prewrite.rs:2192]: Is there a case when the assertion fails on the index key but there is no write conflict now?

Comment (ekexium) [reply src/storage/txn/commands/prewrite.rs:2192]: I can't think of such a case.
We've only encountered 2 false assertion failures, both of which should return write conflicts.

**Review by sticnarf — APPROVED** [review]
LGTM

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**